File: baekjoon_all/01000+/boj_1759.cpp
// Solve 2022-07-12
// Update 2024-02-21

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

string vowels = "aeiou";
int l, c;
char chars[15];
bool chars_checked_vowels[15];
char selected[15];

void dfs(int depth = 0, int begin_idx = 0, int vowel_need = 1, int not_vowel_need = 2) {
    if (depth == l) {
        if (vowel_need <= 0 && not_vowel_need <= 0) {
            for (int i = 0; i < l; i++) {
                cout << selected[i];
            }

            cout << '\n';
        }

        return;
    }

    for (int i = begin_idx; i < c; i++) {
        selected[depth] = chars[i];

        if (chars_checked_vowels[i]) {
            dfs(depth + 1, i + 1, vowel_need - 1, not_vowel_need);
        }
        else {
            dfs(depth + 1, i + 1, vowel_need, not_vowel_need - 1);
        }
    }
}

int main() {
    FASTIO;

    cin >> l >> c;

    for (int i = 0; i < c; i++) {
        cin >> chars[i];
    }

    sort(chars, chars + c);

    for (int i = 0; i < c; i++) {
        for (char v : vowels) {
            if (chars[i] == v) {
                chars_checked_vowels[i] = true;
                break;
            }
        }
    }

    dfs();

    return 0;
}

File: baekjoon_all/01000+/boj_1759.java
// Solve 2024-02-21

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class boj_1759 {

    static char[] vowels = { 'a', 'e', 'i', 'o', 'u' };
    static StringBuilder sb = new StringBuilder();
    static int l, c;
    static char[] chars;
    static boolean[] charsCheckedVowels;
    static char[] selected;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        l = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine(), " ");
        chars = new char[c];
        charsCheckedVowels = new boolean[c];
        selected = new char[l];

        for (int i = 0; i < c; i++) {
            chars[i] = st.nextToken().charAt(0);
        }

        Arrays.sort(chars);

        for (int i = 0; i < c; i++) {
            for (char v : vowels) {
                if (chars[i] == v) {
                    charsCheckedVowels[i] = true;
                    break;
                }
            }
        }

        dfs(0, 0, 1, 2);
        System.out.print(sb);
        br.close();
    }

    static void dfs(int depth, int beginIdx, int vowelNeed, int notVowelNeed) {
        if (depth == l) {
            if (vowelNeed <= 0 && notVowelNeed <= 0) {
                for (char ch : selected) {
                    sb.append(ch);
                }

                sb.append("\n");
            }

            return;
        }

        for (int i = beginIdx; i < c; i++) {
            selected[depth] = chars[i];

            if (charsCheckedVowels[i]) {
                dfs(depth + 1, i + 1, vowelNeed - 1, notVowelNeed);
            } else {
                dfs(depth + 1, i + 1, vowelNeed, notVowelNeed - 1);
            }
        }
    }

}
